### DISCUSSION TITLE
Look for the fastest server of a list and connect to it.

### DISCUSSION THREAD

**Discussion by thaenan** [haugene/docker-transmission-openvpn#1938] [Ideas]
Hello,

I was wondering if it would be possible to screen through a list of servers to try and find the fastest one (lowest ping, lowest load?), then connect to it.

Have any of you already worked on something like it? I was thinking about a shell script to find the fastest one and set the right environment variables. The custom script would then be run before openvpn in the process.

Thanks!

**Comment by pkishino** [comment]
this should go as a suggestion here:
https://github.com/haugene/vpn-configs-contrib
as this is highly provider dependent

**Reply by pkishino** [reply]
@haugene can you please enabled discussions on that repo as well, then we can transfer the discussion there.

**Reply by haugene** [reply]
Will do :+1:

**Comment by haugene** [comment]
Good idea. Lowest ping should be quite easy to script. Another benefit of that is that we would actually ping a bunch of servers before trying to connect so we would avoid trying (and failing) to connect to offline servers.

Lowest load is a tougher nut. I don't think we can know the load before connecting. Some providers might have an API but that would require custom code for each provider. If we were to connect to a bunch of servers and do a bandwidth test for each until we chose the final one, that would take a lot of time. The image is not set up for that anyways. We release control to OpenVPN when we connect to the server, so we would have to change a lot to do something like that.

I moved it to the vpn-configs-contrib repo @pkishino . But now I'm thinking that this probably has to be done in the image repo if anywhere? I'm moving it back for now :see_no_evil:

**Reply by thaenan** [reply]
Hi @haugene,

Lowest ping would definitely be great! 
The load is also considered by my VPN provider (IPVanish) as a criterion for choosing the "Best Server" to connect when using their application, that's why I mentioned it. But as you said, not very easy to get if we have to connect to each and one of the servers of a list. 

Anyway, thank you very much for your interest in the idea and for the great script.

Cheers

**Comment by pkishino** [comment]
Yeah, no, I don’t think this is something we should have in this repo.
Because some providers have api for this and others not etc I think this
should be up to each provider(maintainer) to implement a script for this..
as these things probably will change etc I for sure don’t see this as a
container responsibility..

**Reply by haugene** [reply]
Also a good point. I guess I figured the only way to do this "once and for all" is to do it in the main repo. But this could and maybe should be a per-provider thing.

We do have the `if [[ -x $VPN_PROVIDER_HOME/configure-openvpn.sh ]]; then ...` block in openvpn/start.sh today, so it could actually be implemented in the provider config right now without any changes in the main repo.
